src/lib/replication/repl_dev/raft_repl_dev.cpp

```diff
@@ -1744,4 +1760,113 @@ void RaftReplDev::report_blk_metrics_if_needed(repl_req_ptr_t rreq) {
     }
 }
 
+void RaftReplDev::pause_statemachine() {
+    if (!raft_server()->is_state_machine_execution_paused()) {
+        raft_server()->pause_state_machine_exeuction();
+        while (!raft_server()->wait_for_state_machine_pause(100)) {
+            RD_LOGD(NO_TRACE_ID, "wait for statemachine pause!");
+        }
+    }
+}
+
+void RaftReplDev::resume_statemachine() {
+    if (raft_server()->is_state_machine_execution_paused()) {
+        raft_server()->resume_state_machine_execution();
+        RD_LOGD(NO_TRACE_ID, "statemachine is resumed!");
+    }
+}
+
+void RaftReplDev::enter_emergency() {
+    // all the block allocation happens in rreq->init. so after we wait for all the pending req has been initialized we
+    // can make sure
+    // 1 all the pending reqs has allocated their blocks
+    // 2 no new pending reqs will be initialized again.
+    m_in_emergency.store(true, std::memory_order_release);
+    RD_LOGD(NO_TRACE_ID, "enter emergency state, waiting for all the pending req to be initialized");
+    while (true) {
+        uint64_t pending_req_num = get_pending_init_req_num();
+        if (pending_req_num) {
+            RD_LOGD(NO_TRACE_ID, "wait for {} pending create_req requests to be completed", pending_req_num);
+            std::this_thread::sleep_for(std::chrono::microseconds(1));
+        } else
+            break;
+    }
+}
+
+void RaftReplDev::leave_emergency() { m_in_emergency.store(false, std::memory_order_release); }
```

Comment:
Then we don't need this api, and reset m_is_sm_paused to false in resume_state_machine.